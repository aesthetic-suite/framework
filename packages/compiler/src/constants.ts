import { ScaleType, BreakpointListConfig } from './types';

export const NAME_PATTERN = /^[-a-z0-9]+$/giu;

export const CONFIG_FOLDER = '.aesthetic';

export const BRAND_FILE = 'brand.yaml';

export const LANGUAGE_FILE = 'language.yaml';

export const THEMES_FILE = 'themes.yaml';

export const DEFAULT_BREAKPOINTS: BreakpointListConfig = [640, 960, 1280, 1600, 1920];

export const DEFAULT_UNIT = 8;

export const FONT_FAMILIES = {
  'android-monospace': 'monospace',
  'android-system': 'sans-serif',
  'ios-monospace': 'Menlo-Regular',
  'ios-system': 'SF Pro',
  'web-monospace':
    'Menlo, Monaco, Consolas, "Lucida Console", "Liberation Mono", "Courier New", monospace',
  'web-system':
    'system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"',
};

// https://learnui.design/blog/ultimate-guide-font-sizes-ui-design.html
export const PLATFORM_CONFIGS = {
  android: {
    baseFontSize: 14,
    fontUnit: 'sp',
    spacingUnit: 'dp',
  },
  ios: {
    baseFontSize: 17,
    fontUnit: 'pt',
    spacingUnit: 'pt', // ???
  },
  web: {
    baseFontSize: 16,
    fontUnit: 'px',
    spacingUnit: 'rem',
  },
};

// https://type-scale.com
export const SCALES: { [K in ScaleType]: number } = {
  'augmented-fourth': 1.414,
  'double-octave': 4,
  'golden-ratio': 1.618,
  'golden-section': 1.618,
  'major-second': 1.125,
  'major-third': 1.25,
  'major-sixth': 1.667,
  'major-seventh': 1.875,
  'major-tenth': 2.5,
  'major-eleventh': 2.667,
  'major-twelfth': 3,
  'minor-second': 1.067,
  'minor-third': 1.2,
  'minor-sixth': 1.6,
  'minor-seventh': 1.778,
  octave: 2,
  'perfect-fifth': 1.5,
  'perfect-fourth': 1.333,
};
