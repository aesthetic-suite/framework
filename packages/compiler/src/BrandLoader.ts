import optimal, { string } from 'optimal';
import { DeepPartial } from '@aesthetic/system';
import { Path } from '@boost/common';
import { BRAND_FILE, NAME_PATTERN } from './constants';
import Loader from './Loader';
import { BrandConfigFile } from './types';

export default class BrandLoader extends Loader<BrandConfigFile> {
	getFileName() {
		return BRAND_FILE;
	}

	validate(config: DeepPartial<BrandConfigFile>, filePath: Path): BrandConfigFile {
		return optimal(
			config,
			{
				extends: string(),
				name: string().notEmpty().match(NAME_PATTERN),
			},
			{ file: filePath.path() },
		);
	}
}
