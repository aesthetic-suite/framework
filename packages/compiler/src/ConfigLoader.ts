import fs from 'fs';
import yaml from 'js-yaml';
import { DeepPartial, ColorScheme, ContrastLevel, Hexcode } from '@aesthetic/system';
import optimal, {
  array,
  number,
  object,
  ObjectOf,
  Schema,
  shape,
  string,
  tuple,
  union,
} from 'optimal';
import {
  BorderConfig,
  BorderScaledConfig,
  BorderSizedConfig,
  BreakpointConfig,
  BreakpointListConfig,
  BreakpointSizedConfig,
  ColorConfig,
  ColorShade,
  ConfigFile,
  HeadingScaledConfig,
  HeadingSizedConfig,
  PlatformType,
  ResponsiveConfig,
  Scale,
  ScaleType,
  SpacingConfig,
  SpacingType,
  StrategyType,
  TextScaledConfig,
  TextSizedConfig,
  ThemeConfig,
  TypographyConfig,
} from './types';
import { SCALES, DEFAULT_BREAKPOINTS, DEFAULT_UNIT, FONT_FAMILIES } from './constants';

function hexcode() {
  return string()
    .match(/^#([0-9a-f]{6}|[0-9a-f]{3})$/iu)
    .required();
}

function scale(defaultValue: Scale) {
  return union<Scale>(
    [number().gte(0), string<ScaleType>().oneOf(Object.keys(SCALES) as ScaleType[])],
    defaultValue,
  );
}

function unit(defaultValue: number = 0) {
  return number(defaultValue);
}

function font(platform: PlatformType, type: 'monospace' | 'system') {
  return FONT_FAMILIES[`${platform}-${type}` as 'web-system'];
}

export default class ConfigLoader {
  platform: PlatformType;

  constructor(platform: PlatformType) {
    this.platform = platform;
  }

  async load(filePath: string): Promise<ConfigFile> {
    const data = await fs.promises.readFile(filePath, 'utf8');

    return this.validate(yaml.safeLoad(data));
  }

  validate(config: DeepPartial<ConfigFile>): ConfigFile {
    return optimal(config, {
      borders: this.borders(),
      colors: this.colors(),
      responsive: this.responsive(),
      spacing: this.spacing(),
      themes: this.themes(),
      typography: this.typography(),
    });
  }

  protected borders() {
    const borderScaled = shape<BorderScaledConfig>({
      radius: unit(3),
      radiusScale: scale('perfect-fourth'),
      width: unit(1),
      widthScale: scale(0),
    }).exact();

    const borderSizes = shape<BorderSizedConfig>({
      small: shape({
        radius: unit(2),
        width: unit(1),
      }).exact(),
      default: shape({
        radius: unit(3),
        width: unit(1),
      }).exact(),
      large: shape({
        radius: unit(4),
        width: unit(1),
      }).exact(),
    }).exact();

    return union<BorderConfig>([borderScaled, borderSizes], borderScaled.default());
  }

  protected colors() {
    return array(
      string()
        .notEmpty()
        .camelCase(),
    )
      .notEmpty()
      .required();
  }

  protected responsive() {
    const [xs, sm, md, lg, xl] = DEFAULT_BREAKPOINTS;

    return shape<ResponsiveConfig>({
      strategy: string('mobile-first').oneOf<StrategyType>(['desktop-first', 'mobile-first']),
      breakpoints: union<BreakpointConfig>(
        [
          tuple<BreakpointListConfig>([unit(xs), unit(sm), unit(md), unit(lg), unit(xl)]),
          shape<BreakpointSizedConfig>({
            xs: unit(xs),
            sm: unit(sm),
            md: unit(md),
            lg: unit(lg),
            xl: unit(xl),
          }).exact(),
        ],
        DEFAULT_BREAKPOINTS,
      ),
    }).exact();
  }

  protected spacing() {
    return shape<SpacingConfig>({
      type: string('vertical-rhythm').oneOf<SpacingType>(['unit', 'vertical-rhythm']),
      unit: number(DEFAULT_UNIT),
    }).exact();
  }

  protected themes() {
    return object(
      shape<ThemeConfig>({
        colors: this.themeColors(),
        contrast: string('normal').oneOf<ContrastLevel>(['high', 'low', 'normal']),
        extends: string(),
        palettes: shape({
          brand: this.themePalette(),
          danger: this.themePalette(),
          info: this.themePalette(),
          muted: this.themePalette(),
          neutral: this.themePalette(),
          primary: this.themePalette(),
          secondary: this.themePalette(),
          success: this.themePalette(),
          tertiary: this.themePalette(),
          warning: this.themePalette(),
        })
          .exact()
          .required(),
        scheme: string('light').oneOf<ColorScheme>(['dark', 'light']),
      }).exact(),
    );
  }

  protected themeColors() {
    return object(
      union<Hexcode | ColorConfig>(
        [
          hexcode(),
          shape({
            '00': hexcode(),
            '10': hexcode(),
            '20': hexcode(),
            '30': hexcode(),
            '40': hexcode(),
            '50': hexcode(),
            '60': hexcode(),
            '70': hexcode(),
            '80': hexcode(),
            '90': hexcode(),
          }).exact(),
        ],
        '',
      ),
    )
      .custom(this.validateThemeImplementsColors)
      .required();
  }

  protected themePalette() {
    const color = () => string().custom(this.validatePaletteColorReference);

    const state = () =>
      shape({
        base: color().required(),
        disabled: color(),
        focused: color(),
        hovered: color(),
        selected: color(),
      })
        .exact()
        .required();

    return shape({
      bg: state(),
      fg: state(),
    })
      .exact()
      .required();
  }

  protected typography() {
    return shape<TypographyConfig>({
      font: union<TypographyConfig['font']>(
        [
          string(),
          shape({
            text: string('system'),
            heading: string('system'),
            monospace: string(() => font(this.platform, 'monospace')),
            locale: object(string()),
          }).exact(),
        ],
        'system',
      ),
      text: this.typographyText(),
      heading: this.typographyHeading(),
    }).exact();
  }

  protected typographyHeading() {
    const headingScaled = shape<HeadingScaledConfig>({
      letterSpacing: unit(0.25),
      letterSpacingScale: scale('perfect-fourth'),
      lineHeight: unit(1.5),
      lineHeightScale: scale('major-second'),
      responsiveScale: scale('minor-second'),
      size: unit(16),
      sizeScale: scale('major-third'),
    }).exact();

    const headingSizes = shape<HeadingSizedConfig>({
      level1: shape({
        letterSpacing: unit(1.05),
        lineHeight: unit(2.7),
        size: unit(48),
      }).exact(),
      level2: shape({
        letterSpacing: unit(0.79),
        lineHeight: unit(2.4),
        size: unit(38),
      }).exact(),
      level3: shape({
        letterSpacing: unit(0.59),
        lineHeight: unit(2.14),
        size: unit(32),
      }).exact(),
      level4: shape({
        letterSpacing: unit(0.44),
        lineHeight: unit(1.9),
        size: unit(25),
      }).exact(),
      level5: shape({
        letterSpacing: unit(0.33),
        lineHeight: unit(1.69),
        size: unit(20),
      }).exact(),
      level6: shape({
        letterSpacing: unit(0.25),
        lineHeight: unit(1.5),
        size: unit(16),
      }).exact(),
      responsiveScale: scale('minor-second'),
    }).exact();

    return union<TypographyConfig['heading']>(
      [headingScaled, headingSizes],
      headingScaled.default(),
    );
  }

  protected typographyText() {
    const textScaled = shape<TextScaledConfig>({
      lineHeight: unit(1.25),
      lineHeightScale: scale(0),
      responsiveScale: scale('minor-second'),
      size: unit(14),
      sizeScale: scale('major-second'),
    }).exact();

    const textSizes = shape<TextSizedConfig>({
      small: shape({
        lineHeight: unit(1.25),
        size: unit(12),
      }).exact(),
      default: shape({
        lineHeight: unit(1.25),
        size: unit(14),
      }).exact(),
      large: shape({
        lineHeight: unit(1.25),
        size: unit(16),
      }).exact(),
      responsiveScale: scale('minor-second'),
    }).exact();

    return union<TypographyConfig['text']>([textScaled, textSizes], textScaled.default());
  }

  protected validateThemeImplementsColors = (
    colors: ObjectOf<Hexcode | ColorConfig>,
    schema: Schema<ConfigFile>,
  ) => {
    const names = new Set<string>(schema.struct.colors);
    const unknown = new Set<string>();

    Object.keys(colors).forEach(color => {
      if (names.has(color)) {
        names.delete(color);
      } else {
        unknown.add(color);
      }
    });

    if (names.size > 0) {
      throw new Error(
        `Theme has not implemented the following colors: ${Array.from(names).join(', ')}`,
      );
    }

    if (unknown.size > 0) {
      throw new Error(`Theme is using unknown colors: ${Array.from(unknown).join(', ')}`);
    }
  };

  protected validatePaletteColorReference = (ref: string, schema: Schema<ConfigFile>) => {
    const colors = schema.struct.colors!;

    if (!ref) {
      return;
    }

    // Hue + shade: black.10
    if (ref.includes('.')) {
      const [hue, shade] = ref.split('.') as [string, ColorShade];
      // @ts-ignore TODO need parent object
      const config: string | ColorConfig = colors[hue];

      if (typeof config === 'string') {
        throw new TypeError(
          `Invalid color reference, "${ref}" is pointing to a shade, but the color does not use shades.`,
        );
      } else if (!config || !config[shade]) {
        throw new Error(`Invalid color reference, "${ref}" does not exist.`);
      }

      return;
    }

    // Hue w/ no shade: black
    // @ts-ignore TODO need parent object
    const config: string | ColorConfig = colors[ref];

    if (config && typeof config !== 'string') {
      throw new Error(
        `Invalid color reference, "${ref}" is pointing to shadeless color, but the color is using shades.`,
      );
    } else if (!config) {
      throw new Error(`Invalid color reference, "${ref}" does not exist.`);
    }
  };
}
