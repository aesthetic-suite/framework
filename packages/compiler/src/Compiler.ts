import fs from 'fs-extra';
import optimal, { string } from 'optimal';
import ejs, { TemplateFunction } from 'ejs';
import prettier, { BuiltInParserName } from 'prettier';
import { Path, PortablePath } from '@boost/common';
import { createMixins, DEPTHS } from '@aesthetic/system';
import ConfigLoader from './ConfigLoader';
import SystemDesign from './SystemDesign';
import SystemTheme from './SystemTheme';
import WebPlatform from './platforms/Web';
import { FormatType, SystemOptions, PlatformType, ConfigFile, MixinsTemplate } from './types';
import {
  BORDER_SIZES,
  BREAKPOINT_SIZES,
  HEADING_SIZES,
  PALETTE_TYPES,
  SHADE_RANGES,
  SHADOW_SIZES,
  SPACING_SIZES,
  TEXT_SIZES,
} from './constants';

type Platform = WebPlatform;

const TEMPLATES_FOLDER = Path.resolve('../templates', __dirname);

const PRETTIER_PARSERS: { [ext: string]: BuiltInParserName } = {
  js: 'babel',
  ts: 'typescript',
};

const PRETTIER_IGNORE: { [ext: string]: boolean } = {
  sass: true,
};

export default class Compiler {
  readonly configPath: Path;

  readonly options: Required<SystemOptions>;

  readonly targetPath: Path;

  constructor(configPath: PortablePath, targetPath: PortablePath, options: SystemOptions) {
    this.configPath = this.createAndValidatePath(
      configPath,
      'A configuration file path is required.',
      true,
    );

    this.targetPath = this.createAndValidatePath(
      targetPath,
      'A target destination file path is required.',
    );

    this.options = optimal(options, {
      format: string().oneOf<FormatType>([
        'android',
        'ios',
        'web-cjs',
        'web-css',
        'web-less',
        'web-sass',
        'web-scss',
        'web-js',
        'web-ts',
      ]),
      platform: string().oneOf<PlatformType>(['android', 'ios', 'web']),
    });
  }

  async compile(): Promise<void> {
    // Load and validate the config
    const loader = new ConfigLoader(this.options.platform);
    const { name, themes: themesConfig, ...designConfig } = await loader.load(this.configPath);

    // Create the design system and theme variants
    const design = new SystemDesign(name, designConfig, this.options);
    const themes = this.loadThemes(design, themesConfig);

    // Load the platform
    const platform = this.loadPlatform(design);

    // Write the design system and mixin files
    await this.writeMixinsFile(platform);
    await this.writeDesignFile(design, platform);

    // Write all theme files
    await Promise.all(
      Array.from(themes.entries()).map(([, theme]) => this.writeThemeFile(design, theme, platform)),
    );
  }

  getFormatExtension(): string {
    switch (this.options.format) {
      // TODO
      // case 'android':
      //   return 'java';
      // case 'ios':
      //   return 'swift';
      case 'web-css':
        return 'css';
      case 'web-less':
        return 'less';
      case 'web-sass':
        return 'sass';
      case 'web-scss':
        return 'scss';
      case 'web-ts':
        return 'ts';
      default:
        return 'js';
    }
  }

  getTargetFilePath(fileName: string): Path {
    return this.targetPath.append(`${fileName}.${this.getFormatExtension()}`);
  }

  async loadTemplate(name: string): Promise<TemplateFunction | null> {
    const { format } = this.options;
    const targetFolder = format === 'web-ts' ? 'web-js' : format;
    const templatePath = TEMPLATES_FOLDER.append(targetFolder, `${name}.ejs`);

    // Not all targets use all templates
    if (!templatePath.exists()) {
      return null;
    }

    return ejs.compile(await fs.readFile(templatePath.path(), 'utf8'), {
      filename: templatePath.path(),
    });
  }

  async writeDesignFile(design: SystemDesign, platform: Platform): Promise<void> {
    const template = await this.loadTemplate('design');

    return this.writeFile(
      this.getTargetFilePath(design.name),
      await template!({
        data: design.template,
        design,
        borderSizes: BORDER_SIZES,
        breakpointSizes: BREAKPOINT_SIZES,
        elevationDepths: Object.entries(DEPTHS),
        headingSizes: HEADING_SIZES,
        platform,
        shadowSizes: SHADOW_SIZES,
        spacingSizes: SPACING_SIZES,
        textSizes: TEXT_SIZES,
      }),
    );
  }

  async writeMixinsFile(platform: Platform): Promise<void> {
    const template = await this.loadTemplate('mixins');

    if (!template) {
      return Promise.resolve();
    }

    return this.writeFile(
      this.getTargetFilePath('mixins'),
      await template({
        mixins: this.loadMixins(platform),
        platform,
      }),
    );
  }

  async writeThemeFile(
    design: SystemDesign,
    theme: SystemTheme,
    platform: Platform,
  ): Promise<void> {
    const template = await this.loadTemplate('theme');

    return this.writeFile(
      this.getTargetFilePath(`themes/${theme.name}`),
      await template!({
        data: theme.template,
        design,
        paletteTypes: PALETTE_TYPES,
        platform,
        shadeRanges: SHADE_RANGES,
        theme,
      }),
    );
  }

  protected createAndValidatePath(
    filePath: PortablePath,
    message: string,
    exists: boolean = false,
  ): Path {
    if (!filePath || (typeof filePath !== 'string' && !(filePath instanceof Path))) {
      throw new Error(message);
    }

    const path = Path.resolve(filePath);

    if (exists && !path.exists()) {
      throw new Error(`File path "${path.name()}" does not exist.`);
    }

    return path;
  }

  protected loadMixins(platform: Platform): MixinsTemplate {
    const mixins = createMixins(platform.var);

    return {
      'border-sm': mixins.border.sm,
      'border-df': mixins.border.df,
      'border-lg': mixins.border.lg,
      'box-sm': mixins.box.sm,
      'box-df': mixins.box.df,
      'box-lg': mixins.box.lg,
      'heading-l1': mixins.heading.l1,
      'heading-l2': mixins.heading.l2,
      'heading-l3': mixins.heading.l3,
      'heading-l4': mixins.heading.l4,
      'heading-l5': mixins.heading.l5,
      'heading-l6': mixins.heading.l6,
      'pattern-hidden': mixins.pattern.hidden,
      'pattern-offscreen': mixins.pattern.offscreen,
      'pattern-reset-button': mixins.pattern.reset.button,
      'pattern-reset-input': mixins.pattern.reset.input,
      'pattern-reset-list': mixins.pattern.reset.list,
      'pattern-reset-typography': mixins.pattern.reset.typography,
      'pattern-text-break': mixins.pattern.text.break,
      'pattern-text-truncate': mixins.pattern.text.truncate,
      'pattern-text-wrap': mixins.pattern.text.wrap,
      'shadow-xs': mixins.shadow.xs,
      'shadow-sm': mixins.shadow.sm,
      'shadow-md': mixins.shadow.md,
      'shadow-lg': mixins.shadow.lg,
      'shadow-xl': mixins.shadow.xs,
      'text-sm': mixins.text.sm,
      'text-df': mixins.text.df,
      'text-lg': mixins.text.lg,
    };
  }

  protected loadPlatform(design: SystemDesign): Platform {
    return new WebPlatform(
      this.options.format,
      design.template.text.df.size,
      design.template.spacing.unit,
    );
  }

  protected loadThemes(
    design: SystemDesign,
    themesConfig: ConfigFile['themes'],
  ): Map<string, SystemTheme> {
    const themes = { ...themesConfig };
    const map = new Map<string, SystemTheme>();

    // Load all themes that do not extend another theme
    Object.entries(themes).forEach(([name, config]) => {
      if (!config.extends) {
        map.set(name, design.createTheme(name, config));

        delete themes[name];
      }
    });

    // Load themes that do extend another theme
    Object.entries(themes).forEach(([name, config]) => {
      const { extends: extendsFrom, ...themeConfig } = config;
      const parentTheme = map.get(extendsFrom!);

      // Theme existence is validated before hand
      map.set(name, parentTheme!.extend(name, themeConfig, extendsFrom));
    });

    return map;
  }

  protected async writeFile(filePath: Path, data: string): Promise<void> {
    const ext = filePath.ext(true);
    let contents = data;

    // Make sure the target folder exists
    await fs.ensureDir(filePath.parent().path());

    // Run prettier on the code
    const configPath = await prettier.resolveConfigFile();

    if (configPath && !PRETTIER_IGNORE[ext as 'sass']) {
      const config = (await prettier.resolveConfig(configPath)) || {};

      contents = prettier.format(data, {
        ...config,
        parser: PRETTIER_PARSERS[ext as 'js'] ?? ext,
      });
    }

    // Write the file
    return fs.writeFile(filePath.path(), contents, 'utf8');
  }
}
