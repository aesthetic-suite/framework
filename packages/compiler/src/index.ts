/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Compiler from './Compiler';
import BrandLoader from './BrandLoader';
import LanguageLoader from './LanguageLoader';
import ThemesLoader from './ThemesLoader';

export * from './constants';
export * from './types';

export { Compiler, BrandLoader, LanguageLoader, ThemesLoader };
