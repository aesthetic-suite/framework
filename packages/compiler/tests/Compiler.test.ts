import path from 'path';
import { Compiler } from '../src';

const CONFIG_PATH = path.join(__dirname, '../templates/config.yaml');

describe('Compiler', () => {
  describe('constructor()', () => {
    it('errors if no config path', () => {
      expect(() => new Compiler('', '', { platform: 'web', target: 'web-js' })).toThrow(
        'A configuration file path is required.',
      );
    });

    it('errors if config file is missing', () => {
      expect(() => new Compiler('./foo.yaml', '', { platform: 'web', target: 'web-js' })).toThrow(
        'File path "foo.yaml" does not exist.',
      );
    });

    it('errors if no target path', () => {
      expect(() => new Compiler(CONFIG_PATH, '', { platform: 'web', target: 'web-js' })).toThrow(
        'A target destination file path is required.',
      );
    });

    it('does not error if target path is missing', () => {
      expect(
        () => new Compiler(CONFIG_PATH, './unknown-target', { platform: 'web', target: 'web-js' }),
      ).not.toThrow();
    });

    it('errors for unknown platform', () => {
      expect(
        () =>
          new Compiler(CONFIG_PATH, './unknown-target', {
            // @ts-ignore Allow
            platform: 'osx',
            target: 'web-js',
          }),
      ).toThrow('Invalid field "platform". String must be one of: android, ios, web');
    });

    it('errors for unknown target', () => {
      expect(
        () =>
          new Compiler(CONFIG_PATH, './unknown-target', {
            platform: 'web',
            // @ts-ignore Allow
            target: 'web-tsx',
          }),
      ).toThrow(
        'Invalid field "target". String must be one of: android, ios, web-cjs, web-css, web-less, web-sass, web-scss, web-js, web-ts',
      );
    });
  });
});
