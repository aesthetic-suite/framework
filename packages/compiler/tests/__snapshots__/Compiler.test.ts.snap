// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler compile() compiles all configs 1`] = `
"// Auto-generated by Aesthetic. Do not modify!
import { Design } from '@aesthetic/system';

export default new Design('test', {
  border: {
    sm: {
      radius: '0.14rem', // 2px
      width: '0.04rem', // 0.50px
    },
    df: {
      radius: '0.21rem', // 3px
      width: '0.07rem', // 1px
    },
    lg: {
      radius: '0.32rem', // 4.50px
      width: '0.14rem', // 2px
    },
  },
  breakpoint: {
    xs: {
      query: '(min-width: 45.71em)',
      querySize: 640,
      rootLineHeight: 1.33,
      rootTextSize: '14.94px',
    },
    sm: {
      query: '(min-width: 68.57em)',
      querySize: 960,
      rootLineHeight: 1.42,
      rootTextSize: '15.94px',
    },
    md: {
      query: '(min-width: 91.43em)',
      querySize: 1280,
      rootLineHeight: 1.52,
      rootTextSize: '17.01px',
    },
    lg: {
      query: '(min-width: 114.29em)',
      querySize: 1600,
      rootLineHeight: 1.62,
      rootTextSize: '18.15px',
    },
    xl: {
      query: '(min-width: 137.14em)',
      querySize: 1920,
      rootLineHeight: 1.73,
      rootTextSize: '19.36px',
    },
  },
  heading: {
    l1: {
      letterSpacing: '0.25px',
      lineHeight: 1.5,
      size: '1.14rem', // 16px
    },
    l2: {
      letterSpacing: '0.33px',
      lineHeight: 1.69,
      size: '1.43rem', // 20px
    },
    l3: {
      letterSpacing: '0.44px',
      lineHeight: 1.9,
      size: '1.79rem', // 25px
    },
    l4: {
      letterSpacing: '0.59px',
      lineHeight: 2.14,
      size: '2.23rem', // 31.25px
    },
    l5: {
      letterSpacing: '0.79px',
      lineHeight: 2.4,
      size: '2.79rem', // 39.06px
    },
    l6: {
      letterSpacing: '1.05px',
      lineHeight: 2.7,
      size: '3.49rem', // 48.83px
    },
  },
  shadow: {
    xs: {
      x: '0rem', // 0px
      y: '0.14rem', // 2px
      blur: '0.14rem', // 2px
      spread: '0rem', // 0px
    },
    sm: {
      x: '0rem', // 0px
      y: '0.23rem', // 3.24px
      blur: '0.25rem', // 3.50px
      spread: '0rem', // 0px
    },
    md: {
      x: '0rem', // 0px
      y: '0.37rem', // 5.24px
      blur: '0.44rem', // 6.13px
      spread: '0rem', // 0px
    },
    lg: {
      x: '0rem', // 0px
      y: '0.61rem', // 8.47px
      blur: '0.77rem', // 10.72px
      spread: '0rem', // 0px
    },
    xl: {
      x: '0rem', // 0px
      y: '0.98rem', // 13.71px
      blur: '1.34rem', // 18.76px
      spread: '0rem', // 0px
    },
  },
  spacing: {
    xs: '0.31rem', // 4.38px
    sm: '0.63rem', // 8.75px
    df: '1.25rem', // 17.50px
    md: '2.50rem', // 35px
    lg: '3.75rem', // 52.50px
    xl: '5rem', // 70px
    type: 'vertical-rhythm',
    unit: 17.5,
  },
  text: {
    sm: {
      lineHeight: 1.25,
      size: '0.89rem', // 12.44px
    },
    df: {
      lineHeight: 1.25,
      size: '1rem', // 14px
    },
    lg: {
      lineHeight: 1.25,
      size: '1.13rem', // 15.75px
    },
  },
  typography: {
    font: {
      heading:
        'system-ui, -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"',
      locale: {},
      monospace: '\\"Lucida Console\\", Monaco, monospace',
      text:
        'system-ui, -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"',
      system:
        'system-ui, -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"',
    },
    rootLineHeight: 1.25,
    rootTextSize: '14px',
  },
});
"
`;

exports[`Compiler compile() compiles all configs 2`] = `
"// Auto-generated by Aesthetic. Do not modify!
import design from '..';

export default design.createTheme(
  {
    contrast: 'normal',
    scheme: 'light',
  },
  {
    palette: {
      brand: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      primary: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      secondary: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      tertiary: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      neutral: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      muted: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      danger: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      warning: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      failure: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      success: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
      info: {
        color: {
          '00': '#000',
          '10': '#000',
          '20': '#000',
          '30': '#000',
          '40': '#000',
          '50': '#000',
          '60': '#000',
          '70': '#000',
          '80': '#000',
          '90': '#000',
        },
        bg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
        fg: {
          base: '#000',
          disabled: '#000',
          focused: '#000',
          hovered: '#000',
          selected: '#000',
        },
      },
    },
  },
);
"
`;

exports[`Compiler constructor() errors for unknown target 1`] = `"Invalid field \\"format\\". String must be one of: android, ios, web-css, web-js, web-jsx, web-less, web-sass, web-scss, web-ts, web-tsx"`;
